### PR TITLE
fix dup endpoint for statistic configration

### PR DESCRIPTION
**Pull request by xulingqing**
The watcher writes duplicate VM endpoints in file-sd-config ConfigMap. See the example output, in which case, there are only two VM endpoints (21.0.100.68 and 21.0.100.69), but is written many times in the ConfigMap.

$ kubectl get configmap file-sd-config -n istio-system -o yaml apiVersion: v1 data: staticConfigurations.json: '[{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.68:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.69:15020"]},{"targets":["21.0.100.68:15020"]}]'

### CODE REVIEW

**Comment by douglas-reid** [comment]
@linsun can we review / approve this PR?

Also, what will it take to expand the ownership of this repo?

**Comment by xulingqing** [comment]
> @linsun can we review / approve this PR?
> 
> Also, what will it take to expand the ownership of this repo?

Thanks @Lin  for invite me as owner. Do you have any comment on the pr Doug? If no other comment, I will merge the pr.

**Comment by douglas-reid** [comment]
@xulingqing I'll take a look tomorrow morning.

File: watcher.go [hunk 1]
```diff
@@ -110,9 +112,16 @@ func (w *Watcher) Start(stop <-chan struct{}) {
...
 			default: // add or update
 				log.Printf("handle update workload %s", wle.Spec.Address)
-				newTarget := make(map[string][]string)
-				newTarget["targets"] = append(newTarget["targets"], fmt.Sprintf("%s:15020", wle.Spec.Address))
-				staticConfigurations = append(staticConfigurations, newTarget)
+				newTargetAddr := fmt.Sprintf("%s:15020", wle.Spec.Address)
+
+				// Remove duplicates from the node IPs.
+				if checkDuplicates(staticConfigurations, newTargetAddr) {
+					log.Printf("workload target endpoint %s exists in configuration", newTargetAddr)
```
Comment (douglas-reid) [inline watcher.go]: what is the frequency here?  should we be concerned at all with the logs volume here?  maybe import the istio logging package and use `Debugf` ?

i'd also just break here and save on the `else` indent that follows.

File: watcher.go [hunk 2]
```diff
@@ -175,3 +184,34 @@ func updatePromSDConfigMap(client *kubernetes.Clientset, fileSDConfig *v1.Config
 	}
 	return nil
 }
+
+func checkDuplicates(values []map[string][]string, value string) bool {
```
Comment (douglas-reid) [inline watcher.go]: nit: rename this `isDuplicate` and update the params to be clearer:

func isDuplicate(existing []map[string][]string, newTarget string) bool

File: watcher.go [hunk 3]
```diff
@@ -175,3 +184,34 @@ func updatePromSDConfigMap(client *kubernetes.Clientset, fileSDConfig *v1.Config
...
+		for _, ip := range target["targets"] {
+			if ip == value {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func deDupStaticConfig(values []map[string][]string) []map[string][]string {
```
Comment (douglas-reid) [inline watcher.go]: nit: this deduplicates any configuration right?  should we just name this `dedupConfig` ?

File: watcher.go [hunk 4]
```diff
@@ -175,3 +184,34 @@ func updatePromSDConfigMap(client *kubernetes.Clientset, fileSDConfig *v1.Config
...
+	set := make(map[string]bool)
+	var res []map[string][]string
+
+	for _, target := range values {
+		var flag bool
+		for _, ip := range target["targets"] {
+			if _, v := set[ip]; !v {
+				set[ip] = true
+			} else {
```
Comment (douglas-reid) [inline watcher.go]: nit: you could eliminate the `else` by just `continue` after `set[ip]=true`

File: watcher.go [hunk 5]
```diff
@@ -175,3 +184,34 @@ func updatePromSDConfigMap(client *kubernetes.Clientset, fileSDConfig *v1.Config
...
+	}
+	return false
+}
+
+func deDupStaticConfig(values []map[string][]string) []map[string][]string {
+	set := make(map[string]bool)
+	var res []map[string][]string
+
+	for _, target := range values {
+		var flag bool
```
Comment (douglas-reid) [inline watcher.go:205]: nit: can we call this `duplicate` or similar?

**Review by douglas-reid — COMMENTED** [review]
Thanks for tackling this!

File: watcher.go (hunk 1 above)
Comment (xulingqing) [reply watcher.go]: current log is like the code block when initialize the configmap. Any query after initialization won't generate more log.

```
 $ kubectl -n istio-system logs prometheus-5485f55b86-ksfch -c vm-discovery
2021/05/20 01:02:18 starting new watcher for workload entries
2021/05/20 01:02:18 workload entry watcher created
2021/05/20 01:02:18 waiting to be stopped
2021/05/20 01:02:18 handle update workload 21.0.126.129
2021/05/20 01:02:18 workload 21.0.126.129 already registered as 21.0.126.129:15020
2021/05/20 01:02:18 handle update workload 21.0.125.130
2021/05/20 01:02:18 workload 21.0.125.130 already registered as 21.0.125.130:15020
```

**Pull request merged**